Add style merging

westonganger opened the pull request:
Addresses issue #14

It now only safely patches `Axlsx::Styles.add_style` to implement a style cache. Style index had to be moved from the Workbook to Styles because you cant access the workbook from the `add_styles` method. Tests are all passing but new ones added for this probably aren't super meaningful yet.

sakovias commented:
This is looking very good! I'll try to take a closer look this or next week.

sakovias commented on test/integration_test.rb:
```diff
@@ -229,6 +235,46 @@ def test_mixed_borders
+      sheet.add_style 'B2:D4', b: true
+      sheet.add_style 'B2:D4', border: { style: :thin, color: '000000' }
+    end
+    @workbook.apply_styles
```
Could you replace this line with `serialize(filename)` and replace `@filename` on line 242 with `filename`?

I forgot to remove `@filename` in all places when I was removing tests "teardown" hook that was using the variable. The reason I'm asking to add `serialize` is for visual inspection of test results in the `tmp` folder.

sakovias replied:
I see that both bold font face and border are being applied which is good. Unfortunately, the border is being drawn around every cell. In the original implementation the border is only added to the interior of the selected range (as in the example on the readme).

![border_issue](https://cloud.githubusercontent.com/assets/4750367/18030362/f28871de-6c80-11e6-9ef7-156a40dc07c8.png)

sakovias replied:
Actually, I was wrong on my previous comment:
```
sheet.add_style 'B2:D4', border: { style: :thin, color: '000000' }
```
does exactly what it's supposed to do, which is apply border to every cell in the range. This is how `axlsx` works. The default border is a border applied to every edge of the cell.

sakovias commented on test/integration_test.rb:
```diff
@@ -229,6 +235,46 @@ def test_mixed_borders
+      sheet.add_row ['A2', 'B2']
+      sheet.add_border 'A1:B2'
+    end
+    @workbook.apply_styles
```
Same here please

sakovias replied:
The border is added correctly here, but the bold face on B1 is getting lost.

![bold-issue](https://cloud.githubusercontent.com/assets/4750367/18030375/a4dec93c-6c81-11e6-86b3-71b93f4b92db.png)

sakovias commented on test/integration_test.rb:
```diff
@@ -229,6 +235,46 @@ def test_mixed_borders
+      sheet.add_row ['A2', 'B2']
+      sheet.add_style 'B1:B2', bg_color: 'ff0000'
+    end
+    @workbook.apply_styles
```
And here too please

sakovias replied:
The bold face got lost here as well.

![bold-issue-2](https://cloud.githubusercontent.com/assets/4750367/18030384/075ba562-6c82-11e6-86a3-35825b3278c6.png)

sakovias commented:
I like how clean this change is looking and this seems to be very close to the end goal :smile: 

Lets make sure the new tests that you've added produce the results we want. I'll add some comments in the test file.

sakovias commented on test/cell_test.rb:
```diff
@@ -9,6 +9,6 @@ def test_can_add_style
 
     cell.add_style b: true
-    assert_equal({ b: true }, cell.raw_style)
+    assert_equal({b: true}, cell.raw_style)
```
Very minor nitpick... I think having spaces inside curly braces is a popular Ruby tradition. Rubocop (with default config) likes that too. There are many other Rubocop warnings here though which is all my fault :smile:  (most of this was written before I started using Rubocop).

westonganger commented:
I think the Hash method `key` might not be doing a good job comparing hashes I am going to change this and see what happens.

westonganger commented:
Okay so I fixed the style merging in the latest commit. I had forgotten to add the cell styles from the index to the raw_styles before applying it to the cell. 

As for the borders issue, I noticed that for `test_merge_styles_1` that the problem was fixed when I switched from:
`sheet.add_styles 'B2:D4', borders: { style: :thin, color: '000000' }`
to this:
`sheet.add_border 'B2:D4', { style: :thin, color: '000000' }`

I would appreciate your insight as to why the first statement is failing.

westonganger commented:
Actually I just checked with the current release v0.1.5 and the `sheet.add_styles 'B2:D4', borders: { style: :thin, color: '000000' }` is broken here as well. I suspect that you have incorrect documentation regarding this or `add_styles` need to be fixed to use the BorderCreator.

sakovias commented on lib/axlsx_styler.rb:
```diff
@@ -54,9 +54,12 @@ class Styles
       end
       return index
     end
+
```
Is this just in case or a leftover?

sakovias commented:
I'm really glad you've fixed it! I'll take a closer look some time soon.

I apologise that I was wrong about the border. If we use
```
sheet.add_border 'B2:D4', { style: :thin, color: '000000' }
```
this goes to border creator which only puts up an outer border. But when we use
```
sheet.add_styles 'B2:D4', borders: { style: :thin, color: '000000' }
```
this would apply the given style to every cell in the range, i.e. put border around each cell in `B2:D4`. I think when I was making comments last time I misread `add_styles` for `add_border`. Sorry about that.

sakovias commented on test/integration_test.rb:
```diff
@@ -229,6 +236,51 @@ def test_mixed_borders
+      sheet.add_row ['', '1', '2', '3']
+      sheet.add_row ['', '4', '5', '6']
+      sheet.add_row ['', '7', '8', '9']
+      sheet.add_style 'B2:D4', b: true
```
Could you extract "bold" as a style, so that we test merging old styles and new styles (the same way as you do in the two tests that follow)?
``` ruby
bold = @workbook.styles.add_style b: true
# ...
sheet.add_row ['', '1', '2', '3'], style: [nil, bold, nil, nil]
```

sakovias commented:
I think this is good to go. Good job!

There's one small comment that I've left in the tests (re: testing merge of old and new styles). We can merge when that's cleared.

westonganger commented:
Alright I have added those changes. With these new features maybe you want to bump this gem to `v1.0.0`

sakovias commented on test/test_helper.rb:
```diff
@@ -2,3 +2,5 @@
 require 'minitest/pride'
 require 'awesome_print'
+
+mkdir_p(File.expand_path("../../tmp", __FILE__))
```
Thanks for adding this!

sakovias commented:
Well done! Thank you for your contribution!
